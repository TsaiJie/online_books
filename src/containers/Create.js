import React, { PureComponent } from 'react'
import CategorySelect from '../components/CategorySelect'
import PriceForm from '../components/PriceForm'
import { Tab, Tabs } from '../components/Tabs'
import { testCategories } from '../testData'
import { TYPE_INCOME, TYPE_OUTCOME } from '../utility'
import { AppContext } from '../App'
import withContext from '../WithContext'
class Create extends PureComponent {
  render() {
    const { data } = this.props
    const filterCategories = testCategories.filter(
      (categroy) => categroy.type === TYPE_OUTCOME
    )
    return (
      <div
        className="create-page py-3 px-3 rounded mt-3"
        style={{ background: '#fff' }}
      >
        <Tabs activeIndex={0} onTabChange={() => {}}>
          <Tab>支出</Tab>
          <Tab>收入</Tab>
        </Tabs>
        <CategorySelect
          categories={filterCategories}
          onSelectCategory={() => {}}
        />
        <PriceForm
          onFormSubmit={(a, b) => {
            console.log(a, b)
          }}
          onCancelSubmit={(e) => {
            console.log(e)
          }}
        />
      </div>
    )
  }
}

export default withContext(Create)
