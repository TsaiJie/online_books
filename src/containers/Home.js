import React, { PureComponent } from 'react'
import PriceList from '../components/PriceList'
import Ionicon from 'react-ionicons'
import { AppContext } from '../App'
import {
  LIST_VIEW,
  CHART_VIEW,
  TYPE_INCOME,
  TYPE_OUTCOME,
  parseToYearAndMonth,
  padLeft,
} from '../utility'
import TotalPrice from '../components/TotalPrice'
import MonthPicker from '../components/MonthPicker'
import CreateBtn from '../components/CreateBtn'
import logo from '../logo.svg'
import { Tabs, Tab } from '../components/Tabs'
export const categories = {
  1: {
    id: '1',
    name: '旅行',
    type: 'outcome',
    iconName: 'ios-plane',
  },
  2: {
    id: '2',
    name: '理财',
    type: 'income',
    iconName: 'logo-yen',
  },
}
export const items = [
  {
    id: 1,
    title: '去云南旅游',
    date: '2020-12-10',
    price: 200,
    cid: 1,
  },
  {
    id: 2,
    title: '去云南旅游',
    date: '2020-12-1',
    price: 400,
    cid: 1,
  },
  {
    id: 3,
    title: '理财收入',
    date: '2020-12-22',
    price: 400,
    cid: 2,
  },
]
export const newItem = {
  id: 4,
  title: '新添加的项目',
  date: '2020-12-10',
  price: 300,
  cid: 1,
}
const tabsText = [LIST_VIEW, CHART_VIEW]
export default class Home extends PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      items,
      currentDate: parseToYearAndMonth('2020/12/10'),
      tabView: LIST_VIEW,
    }
  }
  changeView = (index) => {
    this.setState({ tabView: tabsText[index] })
  }
  changeDate = (year, month) => {
    this.setState({ currentDate: { year, month } })
  }
  modifyItem = (modifiedItem) => {
    const modifiedItems = this.state.items.map((item) => {
      if (item.id === modifiedItem.id) {
        return { ...item, title: '更新后的标题' }
      } else {
        return item
      }
    })
    this.setState({
      items: modifiedItems,
    })
  }
  createItem = () => {
    this.setState({
      items: [newItem, ...this.state.items],
    })
  }
  deleteItem = (deletedItem) => {
    const filteredItems = this.state.items.filter(
      (item) => item.id !== deletedItem.id
    )
    this.setState({
      items: filteredItems,
    })
  }

  render() {
    const { items, currentDate, tabView } = this.state
    const itemsWithCategory = items
      .map((item) => {
        item.category = categories[item.cid]
        return item
      })
      .filter((item) => {
        return item.date.includes(
          `${currentDate.year}-${padLeft(currentDate.month)}`
        )
      })
    let totalIncome = 0,
      totalOutcome = 0
    items.forEach((item) => {
      if (item.category.type === TYPE_OUTCOME) {
        totalOutcome += item.price
      } else {
        totalIncome += item.price
      }
    })
    return (
      <AppContext.Consumer>
        {({ state }) => {
          return (
            <React.Fragment>
              <header className="App-header">
                <div className="row mb-5 justify-content-center">
                  <img src={logo} className="App-logo" alt="logo" />
                </div>
                <div className="row">
                  <div className="col">
                    <MonthPicker
                      year={currentDate.year}
                      month={currentDate.month}
                      onChange={this.changeDate}
                    />
                  </div>
                  <div className="col">
                    <TotalPrice income={totalIncome} outcome={totalOutcome} />
                  </div>
                </div>
              </header>
              <div className="content-area py-3 px-3">
                <Tabs activeIndex={0} onTabChange={this.changeView}>
                  <Tab>
                    <Ionicon
                      className="rounded-circle mr-2"
                      fontSize="25px"
                      color={'#007bff'}
                      icon={'ios-paper'}
                    />
                    列表模式
                  </Tab>
                  <Tab>
                    <Ionicon
                      className="rounded-circle mr-2"
                      fontSize="25px"
                      color={'#007bff'}
                      icon={'ios-pie'}
                    />
                    图表模式
                  </Tab>
                </Tabs>
                <CreateBtn onClick={this.createItem} />
                {tabView === LIST_VIEW && (
                  <PriceList
                    items={itemsWithCategory}
                    onModifyItem={this.modifyItem}
                    onDeleteItem={this.deleteItem}
                  />
                )}
                {tabView === CHART_VIEW && (
                  <h1 className="chart-title">这里是图表区域</h1>
                )}
              </div>
            </React.Fragment>
          )
        }}
      </AppContext.Consumer>
    )
  }
}
