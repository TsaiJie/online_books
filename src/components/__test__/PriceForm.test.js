import React from 'react'
import { mount } from 'enzyme'
import PriceFrom from '../PriceForm'
import { testItems } from '../../testData'
import { items } from '../../containers/Home'
const props = {
  onFormSubmit: jest.fn(),
  onCancelSubmit: jest.fn()
}
const propsWithItem = {
  item: testItems[0],
  onFormSubmit: jest.fn(),
  onCancelSubmit: jest.fn()
}
let wrapper, formInstance, wrapper2
export const getInputValue = (selector, wrapper) => (
  wrapper.find(selector).instance().value
)
export const setInputValue = (selector, newValue, wrapper) => (
  wrapper.find(selector).instance().value = newValue
)

describe('test PriceForm component', () => {
  beforeEach(() => {
    wrapper = mount(<PriceFrom {...props} />)
    // 真实节点
    wrapper2 = mount(<PriceFrom {...propsWithItem} />)
    // 节点的实例
    formInstance = wrapper.find(PriceFrom).instance()

  })
  it('should render the component to match snapshot', () => {
    expect(wrapper.getElement()).toMatchSnapshot()
    expect(wrapper2.getElement()).toMatchSnapshot()
  })
  describe('test PriceFrom with no item', () => {
    it('render PriceForm should generate three inputs and one form element', () => {
      expect(wrapper.find('input').length).toEqual(3)
      expect(wrapper.find('form').length).toEqual(1)
    })
    it("render PriceForm with no data should render inputs and no value", () => {
      expect(getInputValue('#title', wrapper)).toEqual('')
      expect(getInputValue('#price', wrapper)).toEqual('')
      expect(getInputValue('#date', wrapper)).toEqual('')
    })
    it('submit form with empty input should show alert message', () => {
      wrapper.find('form').simulate('submit')
      expect(formInstance.state.validatePass).toEqual(false)
      expect(wrapper.find('.alert').length).toEqual(1)
      expect(props.onFormSubmit).not.toHaveBeenCalled()
    })
    it("submit form with invalid price should show alert message", () => {
      setInputValue('#title', 'test', wrapper)
      setInputValue('#price', '-20', wrapper)
      wrapper.find('form').simulate('submit')
      expect(formInstance.state.validatePass).toEqual(false)
    })
    it('submit form with invalid date format should show alert message', () => {
      setInputValue('#title', 'test', wrapper)
      setInputValue('#price', '20', wrapper)
      setInputValue('#date', 'wrong date', wrapper)
      wrapper.find('form').simulate('submit')
      expect(formInstance.state.validatePass).toEqual(false)
    })
    it('submit form with valid data should call with the rigth object', () => {
      setInputValue('#title', 'test', wrapper)
      setInputValue('#price', '20', wrapper)
      setInputValue('#date', '2018-01-01', wrapper)
      const newItem = { title: 'test', price: 20, date: '2018-01-01'}
      wrapper.find('form').simulate('submit')
      expect(props.onFormSubmit).toHaveBeenCalledWith(newItem, false)
    })
    it('click the cancel button should call the right callback', () => {
      wrapper.find('button').last().simulate('click')
      expect(props.onCancelSubmit).toHaveBeenCalled()
    })
    
  })
  describe('test PriceForm with item data', () => {
    it('render PriceForm with item should render the correct data no inputs', () => {
      expect(getInputValue('#title', wrapper2)).toEqual(testItems[0].title)
      expect(getInputValue('#price', wrapper2)).toEqual(testItems[0].price.toString())
      expect(getInputValue('#date', wrapper2)).toEqual(testItems[0].date)
    })
    it('submit with changed value, should trigger callback with right object', () => {
      setInputValue('#price', '200', wrapper2)
      setInputValue('#title', 'new title', wrapper2)
      wrapper2.find('form').simulate('submit')
      const newItem = { ...testItems[0], title: 'new title', price: 200 }
      expect(formInstance.state.validatePass).toEqual(true)
      expect(wrapper2.find('.alert').length).toEqual(0)
      expect(propsWithItem.onFormSubmit).toHaveBeenCalledWith(newItem, true)
    })
  })
})
