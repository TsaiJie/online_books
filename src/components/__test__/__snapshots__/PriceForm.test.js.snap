// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PriceForm component should render the component to match snapshot 1`] = `
<PriceForm
  item={Object {}}
  onCancelSubmit={[MockFunction]}
  onFormSubmit={[MockFunction]}
/>
`;

exports[`test PriceForm component should render the component to match snapshot 2`] = `
<PriceForm
  item={
    Object {
      "cid": "1",
      "date": "2020-12-15",
      "id": "_kly1klf4g",
      "monthCategory": "2020-12",
      "price": 100,
      "timestamp": 1534291200000,
      "title": "buy stuff for kitten",
    }
  }
  onCancelSubmit={[MockFunction]}
  onFormSubmit={[MockFunction]}
/>
`;
